revert some changes: testvectors and re-add lenght prefix

liamsi opened the pull request:
~~If we still want to remove the encoding / len prefixing, here is a branch I've played around with different approaches and where I just kept the changes that fix the tests / build issues~~~

I reverted some changes here: 
- use the original test-vectors from tmlib
- (re)add the encoding / length prefix
- fix the tests

Some documentation:
- document public methods and follow golang documentation best practices
- add a warning (and a TODO) about 2nd preimage attacks (see #107) in a separate file (doc.go)

feel free to merge into your branch (if it might save you some time) or ignore and close 

/cc @ebuchman

liamsi committed:
experimenting with @ebuchman's changes: updated test vectors and other
minor fixes etc

Signed-off-by: Liamsi <Liamsi@users.noreply.github.com>

liamsi committed:
@ebuchman:
meh - trying to simplify the exposed interface and it got replaced with tmhash.Sum

Signed-off-by: Liamsi <Liamsi@users.noreply.github.com>

ebuchman commented:
So now I'm not sure what to do because of https://github.com/tendermint/go-crypto/pull/105#discussion_r190622790 ...

liamsi committed:
revert some changes: add length prefix / basic encoding; test
some minor changes in the documentation (follow golang best practices / golint)

liamsi committed:
commit doc.go

ebuchman reviewed (approved):
Thank you.
